src/normalize.rs

```diff
@@ -0,0 +1,242 @@
+//! Utilities for normalizing BibTeX data
+use std::str::CharIndices;
+
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Default, Deserialize, Serialize)]
+#[serde(deny_unknown_fields)]
+pub struct Normalization {
+    #[serde(default)]
+    normalize_whitespace: bool,
+    #[serde(default)]
+    set_eprint: Vec<String>,
+}
+
+pub trait Normalize {
+    /// Attempt to set the `eprint` and `eprinttype` fields from a given field.
+    ///
+    /// Note that, if successful, this will overwrite the `eprint` and eprinttype` fields.
+    ///
+    /// `eprint` will be set to the corresponding value, and `eprinttype` will be set to the
+    /// corresponding key. Returns `true` if the eprint was set, and `false` otherwise.
+    fn normalize_eprint<Q: AsRef<str>>(&mut self, keys: std::slice::Iter<'_, Q>) -> bool;
+
+    /// Normalize whitespace by converting all whitespace blocks into a single ASCII SPACE,
+    /// respecting whitespace which is explicitly escaped by `\`.
+    fn normalize_whitespace(&mut self) -> bool;
+
+    /// Apply the given normalizations.
+    #[inline]
+    fn normalize(&mut self, nl: &Normalization) {
+        if nl.normalize_whitespace {
+            self.normalize_whitespace();
+        }
+
+        self.normalize_eprint(nl.set_eprint.iter());
+    }
+}
+
+/// Normalize whitespace by converting all blocks of consecutive whitespace into a single ASCII SPACE,
+/// respecting whitespace which is explicitly escaped by `\`.
+///
+/// If the input requires normalization, return the new normalized string. Otherwise, the original
+/// input is already normalized. Note that the returned string, if any, necessarily has a shorter
+/// length than the original string.
+pub fn normalize_whitespace(input: &str) -> Option<String> {
```

Comment:
Can you think of any good names here? Agreed that it could be changed but could not think of something more sensible. I would rather not rename the trait fn (since I think it is best if it is the same as the `config` key)

Reply:
Something like `normalize_whitespace_str`?